Rust support mk2

nrc opened the pull request:
There are a bunch of tests. I haven't worked out the best way to run them. Currently I run them from a shell script (see the last script, they all pass), but I guess I need a makefile or something to setup the env vars (`make tests` doesn't work, but I have Clang problems which I couldn't be bothered to resolve, never mind the Rust stuff).

erikrose commented:
> but I have Clang problems which I couldn't be bothered to resolve

Perhaps you're having the same problems as our CI env suddenly started to have? I'm going to put today toward figuring that out, and maybe you'll benefit from it.

nrc commented:
I suspect, though I haven't tried it has something to do with the `clang` command disappearing in an Ubuntu update (I have to use `clang-3.5` now). I needed to change that for compiling llvm recently, not sure if that is the same problem as affecting DXR though, or affecting your CI.

erikrose commented:
Yep, that's one of them. clang++ disappeared, specifically. I don't know how they can do that in a mere update.

erikrose commented on dxr/filters.py:
```diff
@@ -192,3 +192,44 @@ def highlight_content(self, result):
                 self._should_be_highlit(entity))
+
+
+class QualifiedNameFilterBase(NameFilterBase):
```
Yep, you'll find somebody else has already factored this up.  :-) You might want to merge master into your branch at some point.

erikrose commented on dxr/plugins/rust/__init__.py:
```diff
@@ -0,0 +1,1073 @@
+
+# FIXME(#15) all these process_* methods would be better off in TreeToIndex
+
+""" There should only be one of these per crate and it gives info about the current
```
In Python, docstrings go just *inside* functions (or classes).

erikrose commented on dxr/plugins/rust/tests/test_data/makefile:
```diff
@@ -0,0 +1,12 @@
+	# replaces PWD with `pwd` and produces dxr.config
+	cat dxr.config.in | sed -e 's?PWD?'`pwd`'?g' > dxr.config
+	# Navigate into the DXR folder, build using config file
+	LD_LIBRARY_PATH=$$LD_LIBRARY_PATH:../../trilite PYTHONPATH=$$PYTHONPATH:../../ dxr-build.py
```
We won't need trilite, of course.

erikrose commented on dxr/plugins/rust/tests/test_data/test_data_types.py:
```diff
@@ -0,0 +1,147 @@
+
+class DataTypesTests(DxrInstanceTestCase):
+    def test_struct_name(self):
+        self.found_line_eq('type:NoFields', "struct <b>NoFields</b>;", 6)
```
Looks sane to me! If you were using a SingleFileTestCase, you wouldn't even have to specify the line number. (Those are brittle, obviously, and can be a pain.)

erikrose commented on dxr/plugins/rust/tests/test_data/test_fields.py:
```diff
@@ -0,0 +1,153 @@
+                                           ("EnumFields::Nested(SomeFields { field1, <b>field2</b> }) =&gt; {}", 43)], is_case_sensitive=False)
+
+    def test_qual_field_ref_case_insensitive(self):
+        self.found_lines_eq('+var-ref:test::SomeFields::field1',
```
If you're using the existing QualifiedBlahBlahFilter machinery, I don't know that you need to test case-sensitivity and case-insensitivity, unless you're explicitly doing something to muck with it. That should be already covered for you.

nrc replied:
My thinking was that it is possible to break case senstitivity (e.g., by not using Qualified...), so we should test that it works. I probably went a bit overboard, just one or two tests would be OK, rather than loads.

erikrose commented on dxr/plugins/rust/tests/test_data/code/makefile:
```diff
@@ -0,0 +1,6 @@
+all: code
```
You could use SingleFileTestCase if you didn't want to have all this boilerplate (makefile, dxr.config.in, etc.); we just made the source file name customizable so it doesn't assume "main.cpp". But I assume you're finding it convenient to be able to build and examine in place for now.

erikrose commented on dxr/plugins/rust/tests/test_fns/test_fns.py:
```diff
@@ -0,0 +1,82 @@
+                           is_case_sensitive=False)
+
+    def test_methods_are_functions_too(self):
+        self.found_line_eq('function:baz', "fn <b>baz</b>(&amp;self) {}", 10)
```
It would be nice to comment (or break out into separate tests) to make clear what each of these assertions test: I would guess (1) methods in impls, (2) method declarations (or whatever they're called in Rust) in traits, (3) method definitions in traits.

erikrose commented on dxr/plugins/rust/tests/test_fns/test_fns.py:
```diff
@@ -0,0 +1,82 @@
+
+    def test_method_ref(self):
+        self.found_line_eq('function-ref:baz', "b.<b>baz</b>();", 42)
+        self.found_lines_eq('function-ref:qux',
```
It looks like most of these are referenced from different places and in different ways, so maybe each of these search results tests a distinct code path. If not, though, don't feel shy about breaking things out into tiny, 10-line SingleFileTestCases. Experience has shown me that many small, independent test fixtures are less brittle and more understandable than a few big ones. I leave it to your judgement whether you're at a good place on the continuum. It sure looks like you're thorough, at least! :-D

erikrose commented on dxr/plugins/rust/tests/test_mods/test_fns.py:
```diff
@@ -0,0 +1,55 @@
+from dxr.testing import DxrInstanceTestCase
```
I think you forgot to rename this file.

nrc replied:
No, there are two test files here - one testing the mods themselves, one testing fns within mods.

erikrose commented on dxr/plugins/rust/tests/test_mods/test_fns.py:
```diff
@@ -0,0 +1,55 @@
+from dxr.testing import DxrInstanceTestCase
+
+# These tests are testing references involving modules, as opposed to modules themselves.
```
You can make module-level comments like this into docstrings, and then automated tools will be able to pick them up and do things with them. Just make sure they're the very first thing in the file, even before imports, and put """ around them.

erikrose commented on dxr/plugins/rust/tests/test_mods/test_mods.py:
```diff
@@ -0,0 +1,120 @@
+        self.found_line_eq('+module:test::mod2::mod3', "pub mod <b>mod3</b> {}", 21)
+
+    def test_mod_ref(self):
+        self.found_lines_eq('module-ref:mod1',
```
As above, I wouldn't mind a comment (or a separate test) for each of these assertions so I can understand at a glance what unique thing it's testing.

erikrose commented on dxr/plugins/rust/tests/test_smoke/dxr.config.in:
```diff
@@ -0,0 +1,20 @@
+    [[buglink]]
+    name = github
+    url = https://github.com/mozilla/rust/issues/%s
+    # TODO there needs to be a # at the start, but I have no idea how to make that work
```
Put the whole value in quotes. "#([0-9]+)".

erikrose commented on dxr/plugins/rust/tests/test_smoke/test_smoke.py:
```diff
@@ -0,0 +1,23 @@
+# Index a huge file containing as much of Rust as I could think of. Check that
```
Kinda glad you left this in. :-)

erikrose commented on dxr/plugins/rust/tests/test_smoke/test_smoke.py:
```diff
@@ -0,0 +1,23 @@
+    def test_smoke(self):
+        """Test the index exists and didn't crash."""
+
+        response = self.client().get('/')
```
You don't even need anything here. If it crashes, there'll be an exception, and the explosion will register as a test failure (test error, technically). For comparison, see my test_build_failure.py:

```python
    def test_nothing(self):
        """A null test just to make the setup method run"""
```

erikrose commented on dxr/testing.py:
```diff
@@ -191,8 +191,8 @@ def setup_class(cls):
         chdir(cls._config_dir_path)
-        cls._delete_es_indices()
-        run('make clean')
+        #cls._delete_es_indices()
```
We'll want to undo this before merging, of course.

erikrose commented:
Btw, regarding the clang++ disappearance: that's a deficiency/change in the clang-3.5 package. 3.4 was fine. I had to basically make a symlink in the vagrant provisioning script (actually, I used the Debian alternatives system), and then all was well.

erikrose commented:
This looks really promising. It looks like you're solidly on the right track. I'd of course want to get this tied into the main testrunner, and I'm happy to help you with that. It should be much easier now that the darn tests pass on a fresh Vagrant box (how embarrassing)! We should also get rustc onto the Vagrant box so we can run *all* the tests in CI. How hard is that these days?

Do you feel you're ready to merge? If so, I can schedule this for a deeper review. I didn't really read the plugin itself; I focused more on the tests, per your request. It's very exciting to have Rust support on the horizon!

nrc commented:
Rebased and addressed comments. I couldn't actually figure out how to run DXR any more though, so probably won't actually work.

@erikrose I'll ping you on irc next week. Getting rustc onto vagrant boxes should be easy. Once I get it running again it should be ready to merge, I hope.

erikrose commented:
The new way to run DXR is via a top-level `dxr` command. Those individual dxr-build and dxr-serve commands are no more. For example, you'd go into a test folder and do…

```
dxr index
dxr serve -a
```

I'm sure we can come up with a good way to pass the location of the rust binary in. I'll start thinking about it.

erikrose commented:
Incidentally, a bit more testing boilerplate has gone away. You no longer need a makefile at the top level of each test dir, because…

1. dxr.config now supports relative paths.
2. There's now a `clean_command` option in dxr.config that `dxr clean` runs. (That replaces `make clean` at the top level of the test folder. In fact, `clean_command` defaults to calling `make clean` within the `code` folder.)
3. There's no longer a `target` folder spit out during index, and the temp folder gets automatically deleted, so the clean command doesn't have much left to do.

So you can remove those if you wish; the test harness won't use them anymore.

nrc commented:
So, with the latest rustc (well, a patched version, but give it a day or two), we are passing all tests again. I also removed the makefiles, so I think this PR is about up to date.

I guess the next thing to do is to work out how to include rust and how to add all these tests to the main test suite rather than use the hackey script file. I don't think that should be too much of a problem.

erikrose commented:
I'll have a look tomorrow at integrating the suite.

erikrose commented:
Make that the next day; I'm taking PTO tomorrow. :-)

nrc commented:
@erikrose I'm on PTO all week this week. Thanks for looking into it. My vague 'plan' was to use the rustup script in the vagrant init script:

```
curl -s https://static.rust-lang.org/rustup.sh | sh -s -- --channel=nightly
```

(note that we need the nightly channel).

This will install a global `rustc`. I was planning on changing the environment initialisation stuff to look for a `RUSTC` env var and if it doesn't find it, to use `rustc`. By using the rustup script we get to skip the libs env var stuff, DXR devs will probably still need to set it (if they use a different rustc, which means we can't always just use `rustc`).

erikrose commented:
Cool. I'll hack on it while you're away.

> By using the rustup script we get to skip the libs env var stuff

Do you mean we get to skip setting LD_LIBRARY_PATH for some reason if we use rustup?

nrc commented:
I think so, yes. Of it's not true then we should set it in the vagrant
setup script (I think). Note that if you override the compiler then you
need to override the library path too, so we can't hardwire it in some
where.
On 8 May 2015 10:36 am, "Erik Rose" <notifications@github.com> wrote:

> Cool. I'll hack on it while you're away.
>
> By using the rustup script we get to skip the libs env var stuff
>
> Do you mean we get to skip setting LD_LIBRARY_PATH for some reason if we
> use rustup?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/dxr/pull/393#issuecomment-100039328>.
>

erikrose commented:
I got the rust tests integrated; now `make test` or `nosetests` tests everything. First, I merged master in just to get some other bugfixes, and then I laid down these commits: https://github.com/mozilla/dxr/compare/99d597f03a5a9afeec5d7f7d60c5a84e9df3b7b3...erikrose:nrc-rust5. If that looks good to you, then the next steps are…

1. Understand if we need to do anything to facilitate people overriding LD_LIBRARY_PATH. The tests are working fine on the vagrant box without.
2. Pin to a specific revision of rust so the tests are stable.
3. Merge rust support in! I'm sure there are things we can improve in the plugin code, but that can be an ongoing process, and it'll be awesome to have rust analysis.

Have a look at my commits when you get back and see if they look sane.

erikrose commented:
@nrc, I don't think we'll need that PR#3 from pelmers; I covered that and more in https://github.com/mozilla/dxr/compare/99d597f03a5a9afeec5d7f7d60c5a84e9df3b7b3...erikrose:nrc-rust5 that I mentioned above.

erikrose commented on dxr/build.py:
```diff
@@ -299,9 +299,10 @@ def delete_index_quietly(es, index):
-        rmtree(tree.temp_folder)
+    # TODO undo commenting
+    # if not skip_cleanup:
+    #     # By default, we remove the temp files, because they're huge.
```
No reason not to put this back, right?

nrc replied:
right, sorry, I missed these

erikrose commented on dxr/plugins/buglink/__init__.py:
```diff
@@ -10,6 +10,7 @@
 
 class FileToIndex(dxr.indexers.FileToIndex):
     def refs(self):
+        print "pattern:", self.plugin_config.regex.pattern
```
And I'll remove this.

erikrose commented on dxr/plugins/core.py:
```diff
@@ -225,6 +225,44 @@
+# An unanlyzed string property that points to a value that can be exact- or
+# prefix-matched against and carries start/end bounds for highlighting. Has
+# both a name and a qualname.
+QUALIFIED_NEEDLE = {
```
Note to self: this should move to dxr.indexers.

erikrose commented on dxr/plugins/clang/indexers.py:
```diff
@@ -11,6 +11,7 @@
 from dxr.indexers import (FileToIndex as FileToIndexBase,
                           TreeToIndex as TreeToIndexBase,
                           QUALIFIED_LINE_NEEDLE, unsparsify)
+from dxr.plugins.core import QUALIFIED_NEEDLE
```
Probably unused

nrc replied:
Yeah, looks like a rebasing error or something, sorry

erikrose commented on dxr/query.py:
```diff
@@ -310,6 +309,15 @@ def generic_visit(self, node, visited_children):
         return visited_children or node
 
+# Return the the first filter in filters which is provided by a plugin in enabled_plugins.
+def first_filter_for_plugins(enabled_plugins, filters):
```
Unused?

nrc replied:
Hmm, looks like it. I don't recall if it is meant to be.
